### DISCUSSION TITLE
Docker image for WebCall server?

### DISCUSSION THREAD

**Discussion by mehrvarz** [mehrvarz/webcall#9] [General]
@tonyhkg852 @tna76874

Can you explain to me why you want to run WebCall server with Docker? I would like to understand this better. Are you using Docker to run Nginx? If not, why not?

I've been reading these (and other) articles to catch up:
[Docker: Configuration files](https://medium.com/@jacoelho/docker-configuration-files-60acb1097241)
[Verify repository client with certificates](https://docs.docker.com/engine/security/certificates/)
[Why Docker](https://www.docker.com/why-docker/)

**Comment by tna76874** [comment]
I like to keep my system host system as clean as possible. Every Application is bundeled in a Docker environment with tagged releases. In this way I can manage to administrate a bunch of Applications (including backups) without spending 120% of my time performing updates and stiff on every single application. 🙃

Edit: I run nginx as one of the few host applications (not docker) and use it to manage the reverse proxies to die docker containers. You can also do this with other tools like traefik. Running nginx as a part of a docker-compose bundle is very common. docker-compose is basicly one ore more container that works together to provide a functional app (eg. mysql + anything else)

**Comment by tonyhkg852** [comment]
I am self-hosting a few applications with raspberry pi. The system is also used to test new applications before I consider hosting them, so I'd like to keep things simple and clean. All data and config are kept separately. When there is a need to upgrade the system, it is just a matter of spinning up new containers.

**Comment by mehrvarz** [comment]
I played with Docker a number of years back and never since. I don't really get the point of it. I have no idea what they are talking about on the [Why Docker](https://www.docker.com/why-docker/) page. Please try to convince me. To start with, I have two questions.

@tna76874 you said
> right now there is no persistend config.ini

What do you mean by "persistend" in this context? Adding a config file inside a Docker image does not make much sense. What is a config file worth, if it cannot be changed? And providing a config externally appears to be a major, major hassle. The same goes for access to custom certificate files. Everything seems to become much more complicated. My impression at the moment is that WebCall server would require essential code changes, in order to become a "good" Docker app. Without getting any "better" by that. And oc any change comes with the risk of regression.

I use WebCall as a system service, side by side with Nginx. It can be useful to use Nginx "in front" of WebCall for https. But this is only really true bc Nginx is ususlly serving some unrelated static files also. If there were no static files to serve, and if you were not used to having Nginx on your system, you could just as well use WebCall server without Nginx running at all and it would be just fine. All of WebCall is in a single folder. It is very simple and skinny and only as complicated as it needs to be. Using Docker in between systemd and WebCall looks unnecessary to me (and even scary).

Of course, I could be wrong with all that. Please don't read the above as me trying to convince you. I am only describing where I come from. Instead, try to convince me.

**Reply by tna76874** [reply]
Sorry, typo: persistent

It is no problem to mount specific folders or files from the host system. So you can use config files from outside the container. Also you can map ports from the container to the host. Check out the docker-compose.yml and the Dockerfile in my fork.

**Reply by mehrvarz** [reply]
> It is no problem to mount specific folders or files from the host system.

This must be "Bind mount" in the [Docker: Configuration files](https://medium.com/@jacoelho/docker-configuration-files-60acb1097241) article.

Roughly half of the WebCall config keywords are static. The other half (turnDebugLevel, backupPauseMinutes, logevents, etc.) can be changed at runtime. Do changes to files in the mounted folder make it into the Docker session at runtime? That would great. Maybe no big changes are needed then. 

How far did you get with your fork so far? Do you see any issues coming?

**Reply by tna76874** [reply]
Yes, described there as bind mount. I prefer to structure everything in a docker-compose file. Should be possible to see changes at runtime.

I just put everything together and stopped when there was no reaction when trying to register a new number.

**Reply by mehrvarz** [reply]
> stopped when there was no reaction when trying to register a new number.

wsPort and wssPort are 0 (off) by default. In most cases you only need one of the two. So you need a config file to set wsPort = 8071 (for example) or wssPort = 8443. You can oc enable both. And in case of problems, always check the browser JS console.

**Comment by mehrvarz** [comment]
> I like to keep my system host system as clean as possible.

@tna76874 I wonder what you think of [this](https://timur.mobi/webcall/update/#webcallexecutable), where the executable is the container.
